File: advent_of_code_2023/day11/day11.py
from collections import deque

def arrayise(filename):
    array = []
    f = open(filename, "r")
    for line in f:
        array.append(line.strip())
    return array

def solve(arr):
    universe = []
    for line in arr:
        universe.append([])
        for c in line:
            universe[-1].append(c)
    
    r = 0
    while r < len(universe):
        line = universe[r]
        is_empty = True        
        for c in line:
            if c == '#':
                is_empty = False
                break
        if (is_empty):
            universe.insert(r, ['|' for _ in universe[r]])
            r += 1
        r += 1
    
    c = 0
    while c < len(universe[0]):
        is_empty = True
        for line in universe:
            if line[c] == '#':
                is_empty = False
                break
        if is_empty:
            for line in universe:
                line.insert(c, "|")
            c += 1
        c += 1
    stars = []

    for r, line in enumerate(universe):
        for c, char in enumerate(line):
            if char == '#':
                stars.append((r, c))
    distances = 0
    for i in range(len(stars)):
        star1 = stars[i]
        for j in range(i + 1, len(stars)):
            star2 = stars[j]
            distances += abs(star1[0]-star2[0]) + abs(star1[1]-star2[1])
    print(distances)

    horizontal_distances = []
    vertical_distances = []
    for i, char in enumerate(universe[0]):
        if char == '|':
            horizontal_distances.append(999999)
        else:
            horizontal_distances.append(1)
    for row in universe:
        if row[0] == "|":
            vertical_distances.append(999999)
        else:
            vertical_distances.append(1)
    
    distances = 0
    for i in range(len(stars)):
        star1 = stars[i]
        for j in range(i + 1, len(stars)):
            star2 = stars[j]
            minx = min(star1[0], star2[0])
            maxx = max(star1[0], star2[0])
            miny = min(star1[1], star2[1])
            maxy = max(star1[1], star2[1])
            distances += sum(horizontal_distances[miny:maxy]) + sum(vertical_distances[minx:maxx])
    print(distances)

if __name__ == '__main__':
    filename = "input11.txt"
    arr = arrayise(filename)
    solve(arr)

File: advent_of_code_2023/day11/day11.ts
import {readFileSync} from "fs";

interface Star {
    row: number;
    col: number;
}

function getDistance(s1: Star, s2: Star, 
                     emptyRows: number[], emptyCols: number[], 
                     expansion: number): number {
    
    let minR = Math.min(s1.row, s2.row);
    let maxR = Math.max(s1.row, s2.row);
    let minC = Math.min(s1.col, s2.col);
    let maxC = Math.max(s1.col, s2.col);
    let expansions = emptyRows.filter((v) => minR <= v && v < maxR).length + 
                     emptyCols.filter((v) => minC <= v && v < maxC).length;
    return expansions * (expansion - 1) + (maxR - minR) + (maxC - minC);
}

function solve(lines: string[]): void {
    let emptyRows: number[] = [];
    let emptyCols: number[] = [];
    let isEmptyCols: boolean[] = [];
    let stars: Star[] = [];
    for (let r = 0; r < lines.length; r++) {
        let line = lines[r];
        let isEmptyRow = true;
        for (let c = 0; c < line.length; c++) {
            let char = line[c];
            if (c >= isEmptyCols.length) {
                isEmptyCols.push(true);
            }
            if (char === '#') {
                isEmptyRow = false;
                isEmptyCols[c] = false;
                stars.push({row: r, col: c});
            }
        }
        if (isEmptyRow) {
            emptyRows.push(r);
        }
    }
    emptyCols = isEmptyCols.map((v, i) => v ? i : -1)
                           .filter((x) => x != -1)
    
    let distances1 = 0;
    let distances2 = 0;
    for (let i = 0; i < stars.length; i++) {
        for (let j = i + 1; j < stars.length; j++) {
            let star1 = stars[i];
            let star2 = stars[j];
            distances1 += getDistance(star1, star2, emptyRows, emptyCols, 2);
            distances2 += getDistance(star1, star2, emptyRows, emptyCols, 1000000);
        }
    }
    console.log("Part 1:", distances1);
    console.log("Part 2:", distances2);
}

function main() {
    let lines: string[] = readFileSync("input11.txt", "utf-8")
                            .split(/\n/)
                            .map(x => x.trim())
                            .filter((x => x.length != 0));
    console.time("Time taken");
    solve(lines);
    console.timeEnd("Time taken");
}

main();